mypy/checker.py

```diff
@@ -2744,21 +2744,44 @@ def check_multiple_inheritance(self, typ: TypeInfo) -> None:
         if len(typ.bases) <= 1:
             # No multiple inheritance.
             return
+
         # Verify that inherited attributes are compatible.
-        mro = typ.mro[1:]
-        for i, base in enumerate(mro):
+        # Construct a "typed" MRO that follows regular MRO order, but includes instances
+        # parametrized with their generic args.
+        # This detects e.g. `class A(Mapping[int, str], Iterable[str])` correctly.
+        # For each MRO entry, include it parametrized according to each base inheriting
+        # from it.
+        typed_mro = [
```

Comment:
Looks much better, but I am still thinking, that this `typed_mro` should be builded at MRO calculation phase, not in checker.
This logic should be used shared between all class checks, not compatibility-only, so it will be the code to copy between them, I afraid

Reply:
Probably, we can replace regular MRO to typed one at all? In could be breaking change for some checks, but the idea looks generally correct for me

Reply:
Also, does it deliver types to parent in case, where we specify them no in the final inheritor?

Like in the follow

```python
class A[T]:
    # T = str

class B[T2](A[str]):
     # T2 = int

class C(B[int]):
    ...
```

Reply:
> should be builded at MRO calculation phase, not in checker.

Makes sense to me, but I'd rather defer this to a follow-up PR. There are several more places (override checks) where a "typed MRO" would be handy, we should adjust them one by one - that'd make this PR a lot harder to review. 

> Probably, we can replace regular MRO to typed one at all?

Those are of different types. Fundamentally, `TypeInfo` and `Instance` are different entities that shouldn't be mixed up. It makes sense to me as a general idea, but now it would result in a very huge changeset I won't be able to handle correctly (and no one will enjoy reviewing). `mypy

> does it deliver types to parent correctly

Hope so, `map_instance_to_supertype` looks pretty robust. If I understand your snippet correctly, it's similar to `testMultipleInheritanceCompatibleTypeVar` - generics are mapped there as expected.

Reply:
> Makes sense to me, but I'd rather defer this to a follow-up PR. There are several more places (override checks) where a "typed MRO" would be handy, we should adjust them one by one - that'd make this PR a lot harder to review.

So, can we make it a `@property` in TypeInfo? This case we can do not duplicate the logic in various checkers and have a migration guide from `mro` to `typed_mro`. Or even make `typed_mro` not property, but static calculated attribute in the future.

Also, can you show me the places require `typed_mro` as well. I would like to help with migration. Or, probably, can I help any other way?

Reply:
Not sure if `property` is good (current codebase seems to follow "only trivial computations for properties" rule). A method would be more reasonable, but the problem is in dependencies between those files (note that `mypy.nodes` never imports `mypy.types` - the dependency is inverted, and that was done for architectural reasons). TBH I think this should be a method of `TypeChecker` - does that sound good enough?

Reply:
Regarding other places - I have a strong feeling that other override-related issues would benefit from using typed MRO explicitly. 

However, I looked through the issue tracker now, and seems like almost everything else is fine - the only problem I see right now is #6184 which can indeed be solved using "typed MRO" here:

https://github.com/python/mypy/blob/14974072c0a70f8ca29c17c740475187b800e714/mypy/checker.py#L1983-L1993

But you can just search for `.mro[1:]` in `checker.py`, some usages of it may be problematic. `check_compatibility_all_supers` should also be using typed MRO if I understand it correctly. Any `final` checks, OTOH, don't need that and work with plain MRO just as good.

Reply:
> TBH I think this should be a method of `TypeChecker` - does that sound good enough?

LGTM 👍🏿

Reply:
I can try to take care about #6184 if you are okay with it

Reply:
Thanks for your review! I extracted a helper method (and also stopped duplicating non-generic bases in typed MRO) and borrowed a testcase.

Sure, go ahead with that ticket, that would be great! Please also try to search the issue tracker, I might have missed other reports related to that one.